src/xdp/apikernel.c

```diff
@@ -0,0 +1,539 @@
+    _Out_writes_bytes_(*OutputLength) VOID *OutputValue,
+    _Inout_ UINT32 *OutputLength
+    )
+{
+    UNREFERENCED_PARAMETER(Socket);
+    UNREFERENCED_PARAMETER(OptionName);
+    UNREFERENCED_PARAMETER(InputValue);
+    UNREFERENCED_PARAMETER(InputLength);
+    UNREFERENCED_PARAMETER(OutputValue);
+    UNREFERENCED_PARAMETER(OutputLength);
+    return STATUS_NOT_SUPPORTED;
+}
+
+static
+VOID *
+XdpApiKernelXdpGetRoutine(
+    _In_z_ const CHAR *RoutineName
+    )
+{
+    for (UINT32 i = 0; i < RTL_NUMBER_OF(XdpApiRoutines); i++) {
+        if (strcmp(XdpApiRoutines[i].RoutineName, RoutineName) == 0) {
+            return XdpApiRoutines[i].Routine;
+        }
+    }
+
+    return NULL;
+}
+
+static
+XDP_STATUS
+XdpApiKernelXdpRssGetCapabilities(
+    _In_ HANDLE InterfaceHandle,
+    _Out_writes_bytes_opt_(*RssCapabilitiesSize) XDP_RSS_CAPABILITIES *RssCapabilities,
+    _Inout_ UINT32 *RssCapabilitiesSize
+    )
+{
+    ASSERT(!((RssCapabilities != NULL) ^ (*RssCapabilitiesSize > 0)));
+    return XdpInterfaceOffloadRssGetCapabilities((XDP_INTERFACE_OBJECT *)InterfaceHandle, RssCapabilities, *RssCapabilitiesSize, RssCapabilitiesSize);
+}
+
+static
+XDP_STATUS
+XdpApiKernelXdpRssSet(
+    _In_ HANDLE InterfaceHandle,
+    _In_ const XDP_RSS_CONFIGURATION *RssConfiguration,
+    _In_ UINT32 RssConfigurationSize
+    )
+{
+    return XdpInterfaceOffloadRssSet((XDP_INTERFACE_OBJECT *)InterfaceHandle, RssConfiguration, RssConfigurationSize);
+}
+
+static
+XDP_STATUS
+XdpApiKernelXdpRssGet(
+    _In_ HANDLE InterfaceHandle,
+    _Out_writes_bytes_opt_(*RssConfigurationSize) XDP_RSS_CONFIGURATION *RssConfiguration,
+    _Inout_ UINT32 *RssConfigurationSize
+    )
+{
+    ASSERT(!((RssConfiguration != NULL) ^ (*RssConfigurationSize > 0)));
+    return XdpInterfaceOffloadRssGet((XDP_INTERFACE_OBJECT *)InterfaceHandle, RssConfiguration, *RssConfigurationSize, RssConfigurationSize);
+}
+
+static
+XDP_STATUS
+XdpApiKernelXdpQeoSet(
+    _In_ HANDLE InterfaceHandle,
+    _Inout_ XDP_QUIC_CONNECTION *QuicConnections,
+    _In_ UINT32 QuicConnectionsSize
+    )
+{
+    UINT32 OutputBufferLength = QuicConnectionsSize;
+
+    return XdpInterfaceOffloadQeoSet((XDP_INTERFACE_OBJECT *)InterfaceHandle, QuicConnections, QuicConnectionsSize, &OutputBufferLength);
+}
+
+static
+NTSTATUS
+XdpApiKernelProviderAttachClient(
+    _In_ HANDLE NmrBindingHandle,
+    _In_ const VOID *ProviderContext,
+    _In_ const NPI_REGISTRATION_INSTANCE *ClientRegistrationInstance,
+    _In_ const VOID *ClientBindingContext,
+    _In_ const VOID *ClientNpiDispatch,
+    _Outptr_ VOID **ProviderBindingContext,
+    _Outptr_result_maybenull_ const VOID **ProviderDispatch
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    XDPAPI_PROVIDER *Provider = (XDPAPI_PROVIDER *)ProviderContext;
+    XDPAPI_CLIENT *Client = NULL;
+    const XDP_API_CLIENT_DISPATCH *ClientDispatch = ClientNpiDispatch;
+
+    TraceEnter(TRACE_CORE, "ProviderContext=%p", ProviderContext);
```

Comment:
in my experience, we need some way to uniquely identify "who" attached.

Reply:
what is "who" in this context? application process?

Reply:
@mtfriesen are you just saying you want to also log `ClientBindingContext`?